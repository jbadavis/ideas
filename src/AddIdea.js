import React from 'react';

import styles from './AddIdea.module.css';

const MAX_CHARS = 140;

class AddIdea extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      title: this.props.title,
      description: this.props.description,
      count: this.props.description.length,
      overCount: false,
    };
  }

  handleTitleChange(e) {
    const value = e.target.value;

    this.setState(prevState => ({
      ...prevState,
      title: value,
    }));
  }

  handleDescriptionChange(e) {
    const value = e.target.value;
    const charCount = value.length;

    this.setState(prevState => ({
      ...prevState,
      ...(charCount > MAX_CHARS
        ? {
            overCount: true,
          }
        : {
            description: value,
            count: value.length,
            overCount: false,
          }),
    }));
  }

  handleSubmit = event => {
    event.preventDefault();

    this.props.handleUpdateIdea(this.state.title, this.state.description);
  };

  render() {
    return (
      <>
        <h2>Edit</h2>
        <form className={styles.addIdea} onSubmit={this.handleSubmit}>
          <label>
            Name
            <input
              className={styles.title}
              autoFocus
              type="text"
              name="title"
              onChange={e => this.handleTitleChange(e)}
              value={this.state.title}
            />
          </label>
          <label>
            Description
            <textarea
              className={styles.description}
              name="description"
              onChange={e => this.handleDescriptionChange(e)}
              value={this.state.description}
            />
          </label>
          <div className={styles.charCount}>
            {this.state.count}/{MAX_CHARS}
          </div>
          <input className="submit" type="submit" value="Save" />
        </form>
      </>
    );
  }
}

export default AddIdea;
