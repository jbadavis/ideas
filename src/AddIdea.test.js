import React from 'react';
import AddIdea from './AddIdea';
import {mount, shallow} from 'enzyme';

describe('Idea', () => {
  const titleMock = 'title';
  const descriptionMock = 'description';
  const handleUpdateIdeaMock = jest.fn();

  const component = mount(
    <AddIdea
      title={titleMock}
      description={descriptionMock}
      handleUpdateIdea={handleUpdateIdeaMock}
    />,
  );

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('if the submit CTA is clicked', () => {
    it('should call handleEditIdea', () => {
      component.find('form').simulate('submit');

      expect(handleUpdateIdeaMock).toHaveBeenCalledWith(
        titleMock,
        descriptionMock,
      );
    });
  });
});
