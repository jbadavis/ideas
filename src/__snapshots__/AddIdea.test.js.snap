// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Idea should match snapshot 1`] = `
<AddIdea
  description="description"
  handleUpdateIdea={[MockFunction]}
  title="title"
>
  <h2>
    Edit
  </h2>
  <form
    className="addIdea"
    onSubmit={[Function]}
  >
    <label>
      Name
      <input
        autoFocus={true}
        className="title"
        name="title"
        onChange={[Function]}
        type="text"
        value="title"
      />
    </label>
    <label>
      Description
      <textarea
        className="description"
        name="description"
        onChange={[Function]}
        value="description"
      />
    </label>
    <div
      className="charCount"
    >
      11
      /
      140
    </div>
    <input
      className="submit"
      type="submit"
      value="Save"
    />
  </form>
</AddIdea>
`;
